tools/lsp/main.rs

```diff
@@ -302,33 +312,41 @@ fn main_loop(connection: Connection, init_param: InitializeParams, cli_args: Cli
         Poll::Pending => futures.push(first_future),
     };
 
-    for msg in &connection.receiver {
-        match msg {
-            Message::Request(req) => {
-                // ignore errors when shutdown
-                if connection.handle_shutdown(&req).unwrap_or(false) {
-                    return Ok(());
-                }
-                futures.push(Box::pin(rh.handle_request(req, &ctx)));
-            }
-            Message::Response(resp) => {
-                if let Some(q) = request_queue.lock().unwrap().get_mut(&resp.id) {
-                    match q {
-                        OutgoingRequest::Done(_) => {
-                            return Err("Response to unknown request".into())
+    loop {
+        crossbeam_channel::select! {
+            recv(connection.receiver) -> msg => {
+                match msg? {
```

Comment:
This is a change of behavior as we previously did not error in case of an error here. 
But I guess this is good. We probably should error out since it is not a graceful shutdown.